File: tree/563-binary-tree-tilt.py
# Solution 1: iterative solution using
# stack and hash table
# O(N) in time because we visit each node once
# space complexity of sum_ dictionary: O(N)
# space complexity of stack:
# - O(logN) in average because the size of the stack
#   is the depth of current node.
# - O(N) in worst case when the tree is skewed
# => space complexity overall: O(N)
class Solution:
    def findTilt(self, root):
        # Base case
        if not root:
            return 0

        stack = [root]
        sum_ = {}
        sum_[None] = 0
        tilt = 0
        while stack:
            node = stack[-1]
            if (not node.left or node.left in sum_) and \
               (not node.right or node.right in sum_):
                # if node's children are already visited
                tilt += abs(sum_[node.left] - sum_[node.right])
                sum_[node] = node.val + sum_[node.left] + sum_[node.right]
                stack.pop()
            else:
                if node.left:
                    stack.append(node.left)
                if node.right:
                    stack.append(node.right)
        return tilt

File: practices/leetcode/tree/563-binary-tree-tilt.cpp
// Solution 1: recursion inside recursion
// N = number of nodes in tree
// O(NlogN) in time in average
// Proof: by solving T1, T2
// O(N) in space if skewed tree
// O(logN) in space in average
class Solution {
public:
    // T2(n) = n
    int getSum(TreeNode* node) {
        if (node == NULL) return 0;
        return node->val + getSum(node->left) + getSum(node->right);
    }

    // T1(n) = 2T2(n/2) + 2T1(n/2)
    int findTilt(TreeNode* root) {
        // Base case
        if (root == NULL) {
            return 0;
        }

        int sumLeft = getSum(root->left);
        int sumRight = getSum(root->right);

        return abs(sumLeft - sumRight) +
               findTilt(root->left) +
               findTilt(root->right);
    }
};

// Solution 2: 1 recursion
// Instead of compute sum of each node while traversing,
// we can pre-compute them and save the sum in a hash map.
// O(2N) = O(N) in time
// O(N) in space in average
class Solution {
public:
    int getSum(TreeNode* node, unordered_map<TreeNode*, int> & sum) {
        if (node == NULL) {
            return sum[node] = 0;
        }

        return sum[node] = node->val +
                           getSum(node->left, sum) +
                           getSum(node->right, sum);
    }

    int findTilt_(TreeNode* node, unordered_map<TreeNode*, int> & sum) {
        if (node == NULL) {
            return 0;
        }

        return abs(sum[node->left] - sum[node->right]) +
               findTilt_(node->left, sum) +
               findTilt_(node->right, sum);
    }

    int findTilt(TreeNode* root) {
        unordered_map<TreeNode*, int> sum;
        getSum(root, sum);  // pre-calculate sum of tree at each node in O(N) time

        return findTilt_(root, sum);  // calculate the tilt of each node in O(N) time
    }
};

// Solution 3: optimal recursion
// Computing the tile of each node while computing
// the sum
// O(N) in time
// O(N) in space if skewed tree
// O(logN) in space in average
// pattern: post-order traverse
class Solution {
public:
    int getSum(TreeNode* node, int & tilt) {
        if (node == NULL) return 0;

        int sumLeft = getSum(node->left, tilt);
        int sumRight = getSum(node->right, tilt);

        tilt += abs(sumLeft - sumRight);

        return node->val + sumLeft + sumRight;
    }

    int findTilt(TreeNode* root) {
        int tilt = 0;
        getSum(root, tilt);
        return tilt;
    }
};

// Solution 4: iterative solution
// Please refer to Python version, solution 1.
